lib/ecto/changeset.ex

```diff
@@ -1392,6 +1393,72 @@ defmodule Ecto.Changeset do
     end
   end
 
+  @doc """
+  Validates (tentatively) that no existing record with a different primary key
+  has the same values for these fields.
+
+  This is tentative because a race condition may occur. A unique constraint
+  should also be used to ensure uniqueness.
+
+  However, in most cases where conflicting data exists, it will have been
+  inserted prior to the current validation phase. Noticing those conflicts
+  during validation gives the user a chance to correct them at the same time
+  as other validation errors.
+
+  ## Examples
+      validate_unique_tentatively(changeset, [:email], repo)
+      validate_unique_tentatively(changeset, [:city_name, :state_name], repo)
+      validate_unique_tentatively(changeset, [:city_name, :state_name], repo, "city must be unique within state")
+
+  """
+  def validate_unique_tentatively(changeset, field_names, repo, error_message \\ "has already been taken") do
+    field_names = List.wrap(field_names)
+    where_clause = Enum.map(field_names, fn (field_name) ->
+      {field_name, get_field(changeset, field_name)}
+    end)
+
+    # If we don't have values for all fields, we can't query for uniqueness
+    if Enum.any?(where_clause, fn (tuple) -> is_nil(elem(tuple, 1)) end) do
+      changeset
```

Comment:
No worries, I have added this to master and also renamed it to unsafe_validate_unique. :)